Lowering rendering resolution option

mar753 opened the pull request:
When I was testing my additional pull request implementing RTT: https://github.com/reicast/reicast-emulator/pull/1460 I saw that Adreno 506 has performance issues when rendering using native resolution (1920x1080). I have observed framerate drops in THPS2, so I came with an idea that we should be able to lower the rendering resolution, thus improve the performance.

I have implemented a solution where we can decrease vertical and horizontal resolution (%) in emulator experimental options section.
How it works?
Instead of rendering to the default framebuffer, we render the whole screen to a texture (with native resolution size) but we set a lower viewport value (depending on vertical/horizontal resolution options). Then we use this texture and render it to the default framebuffer, extending it to cover the full screen. Then we apply scissor settings and draw OSD (it is drawn at the full resolution to keep the highest quality). When vertical/horizontal resolutions are set to 100% (default value), the current solution (what is in master branch) is used - so there is no impact on the performance.

Here is an example when full native resolution is used (currently in master branch), vertical/horizontal resolution set to 100% (1920x1080):
![screenshot_2019-01-07-22-12-01-060_com reicast emulator](https://user-images.githubusercontent.com/5397997/50794616-e586f980-12cb-11e9-9633-ab80bc3b5dd7.png)

And here is the example where vertical and horizontal resolutions were set to 25% of the native screen resolution (about: 480x270):
![screenshot_2019-01-07-22-10-37-140_com reicast emulator](https://user-images.githubusercontent.com/5397997/50794775-60501480-12cc-11e9-92bb-9959c40fb064.png)

If you would like to test how this solution works, please use my branch: 
https://github.com/mar753/reicast-emulator/tree/lowering_rendering_resolution_option_private
because current master doesn't work on my Xiaomi Pocophone/Redmi 4 Note devices (works only on Huawei P8 lite) - when I try to load a game I am entering the bios settings.

Please review my second PR as well, where multipass render to texture was implemented: https://github.com/reicast/reicast-emulator/pull/1460
Could you create a test branch for me in this repository as well and connect it to CI?

mar753 committed:
Changing screen resolution: Rendering quad created

mar753 committed:
Rendering resolution lowering (horizontal and vertical) is possible from the configuration menu

mar753 committed:
Additional improvements to lowering rendering resolution feature

mar753 committed:
Merge branch 'master' of https://github.com/reicast/reicast-emulator into lowering_rendering_resolution_option

baka0815 reviewed (changes requested):
Looks mostly good to me, however I don't now much about OpenGL, maybe @flyinghead can help on that end?

But you would also want to make the setting accessible to other OSes and not only Android, so add a setting to the emu.cfg. See https://github.com/reicast/reicast-emulator/blob/6936bc2ec503594e9fc75e687fa060426bb240a5/core/nullDC.cpp#L284

mar753 committed:
nullDC.cpp updated to support other OSes

mar753 commented:
@baka0815 nullDC.cpp file was updated

baka0815 commented:
Thanks @mar753 looks good to me then.

However as stated above I would like to have someone review this with some knowledge in OpenGL like @flyinghead.
Any comments @dmiller423?

flyinghead commented:
This would be a useful feature but do we really need separate horizontal and vertical settings? I think a single "Render Scale Factor"  applied in both directions would be enough.

I tested this on linux with an NVidia 1080 and I get wrong colors when scaling is on:
![mar753-crazy-taxi-render-scale](https://user-images.githubusercontent.com/16261910/51084111-19777a00-1725-11e9-9466-04a8efbe99ce.png)

mar753 committed:
Stencil support for OpenGL added - requires OpenGL >= 3

mar753 committed:
Single render scale factor seekbar

mar753 commented:
This feature was added mainly due to mobile platforms, where performance may be lower. Desktop implementations do not need to lower the resolution. There is no stencil attachment defined so this issue in Crazy Taxi is visible. I have added support for GL_DEPTH24_STENCIL8 in case of OpenGL rendering, but we need to remember that this requires OpenGL >= 3.0 version. This should fix the issue.

I thought about single seekbar for vertical/horizontal resolution, but having them separate you can control the aspect ratio more precisely. Anyway, maybe it is not so important - I have changed it to a single seekbar.

Waiting for feedback

baka0815 commented on shell/android-studio/reicast/src/main/java/com/reicast/emulator/config/OptionsFragment.java:
```diff
@@ -411,47 +411,26 @@ public boolean onEditorAction(TextView v, int actionId, KeyEvent event) {
 			public void onStopTrackingTouch(SeekBar seekBar) {
 				int progress = seekBar.getProgress();
+				//the same progress for both resolutions
 				mPrefs.edit().putInt(Emulator.pref_resolutionv, progress + 1).apply();
```
As we now have only one setting (on Android) we should also only have one setting in the ``emu.cfg`` as well.

baka0815 reviewed (changes requested):
I would like to see the same setting change for the ``emu.cfg`` as we now only have one setting to change (just call it "rend.ResolutionPercentage" or something like that).

Trying to run Crazy Taxi on Linux (Radeon RX 580, mesa 18.2.8, radeonsi) I'm getting the following error and reicast just dies:
````
GL context version: 4.4
OSD: 6
Shader: failed to compile 
0:1(1): error: syntax error, unexpected NEW_IDENTIFIER

Shader linking: failed to link 
 (51 bytes), - error: linking with uncompiled/unspecialized shader -
Fatal error : shader compile fail

 in gl_CompileAndLink -> /home/baka/svn/reicast-master/shell/linux/../../core/rend/gles/gles.cpp : 895 
````

mar753 committed:
Additional improvements/fixes

mar753 committed:
LRR: Support for OpenGL 2 added

mar753 commented:
It turns out that Nvidia GPUs are much more lenient in case of GLSL than Ati GPUs. I have updated vertex and fragment shaders and tested with glslangValidator with every OpenGL and OpenGL ES GLSL version. Shaders compile without any warnings from now on.

I have removed old options from nullDC.cpp and added "rend.ResolutionPercentage" setting as requested.

From now on, when shader compilation fails, we will know if the problem was caused by vertex or fragment shader.

I have added a support for OpenGL 2 GPUs as well, but there will not be a stencil renderbuffer supported when resolution will be lowered (anyway, I doubt that anyone is using so old graphic cards for emulation these days).

mar753 commented:
It turns out that Nvidia GPUs are much more lenient in case of GLSL than Amd/Ati Radeon GPUs. I have updated vertex and fragment shaders and tested with glslangValidator with every OpenGL and OpenGL ES GLSL version. Shaders compile without any warnings from now on.

I have removed old options from nullDC.cpp and added "rend.ResolutionPercentage" setting as requested.

From now on, when shader compilation fails, we will know if the problem was caused by vertex or fragment shader.

I have added a support for OpenGL 2 GPUs as well, but there will not be a stencil renderbuffer supported when resolution will be lowered (anyway, I doubt that anyone is using so old graphic cards for emulation these days).

mar753 commented:
As I have checked FBOs are not supported in OpenGL 2.x without the "GL_EXT_framebuffer_object" extension, so this last commit is not needed because OpenGL 3.0 at least is required anyway.

mar753 committed:
Revert "LRR: Support for OpenGL 2 added"

This reverts commit c33c0cade8dbf8459b881500b1ebd9723e659073.

spielvan commented:
> It turns out that Nvidia GPUs are much more lenient in case of GLSL than Amd/Ati Radeon GPUs. I have updated vertex and fragment shaders and tested with glslangValidator with every OpenGL and OpenGL ES GLSL version. Shaders compile without any warnings from now on.
> 
> I have removed old options from nullDC.cpp and added "rend.ResolutionPercentage" setting as requested.
> 
> From now on, when shader compilation fails, we will know if the problem was caused by vertex or fragment shader.
> 
> I have added a support for OpenGL 2 GPUs as well, but there will not be a stencil renderbuffer supported when resolution will be lowered (anyway, I doubt that anyone is using so old graphic cards for emulation these days).

My Android tv box still uses Open Gl 2.0 ES with a Mali 450, I do not know why this bug occurs that looks like the image has been cropped in the middle of the losangulo.

[https://3.bp.blogspot.com/-ZghEWYDeWwk/XEU2hsEh8VI/AAAAAAAADXY/rYZhAOs-064GRR6X68Tpn0XDHCRZb57kQCLcBGAs/s1600/Screenshot_2019-01-20-23-20-02.png](https://3.bp.blogspot.com/-ZghEWYDeWwk/XEU2hsEh8VI/AAAAAAAADXY/rYZhAOs-064GRR6X68Tpn0XDHCRZb57kQCLcBGAs/s1600/Screenshot_2019-01-20-23-20-02.png)

mar753 commented:
Ok, maybe it was confusing but I was writing about desktop OpenGL 2.0, NOT OpenGL ES 2.0 (where FBOs are available of course).

Hard to say why this issue on your TV appears. Have you lowered the resolution?
I was testing with my Huawei P8 lite with the same GPU as you have and I have not observed any issues. This may depend on aspect ratio or your hardware, try to use different % values for resolution.
BTW I saw much more issues with this GPU which are/were not handled properly, like the one with glReadPixels(), where the only read format handled by this GPU is RGB565, what was causing artifacts in e.g. Rez (using ARGB1555) - this one I have fixed. There is one more issue in THPS2 when you try to exit to the main menu while playing - black screen appears (instead of "Loading...", like on Adreno GPUs).
Anyway, these issues are not related to this pull request or the #1460.

spielvan commented:
> Ok, maybe it was confusing but I was writing about desktop OpenGL 2.0, NOT OpenGL ES 2.0 (where FBOs are available of course).
> 
> Hard to say why this issue on your TV appears. Have you lowered the resolution?
> I was testing with my Huawei P8 lite with the same GPU as you have and I have not observed any issues. This may depend on aspect ratio or your hardware, try to use different % values for resolution.
> BTW I saw much more issues with this GPU which are/were not handled properly, like the one with glReadPixels(), where the only read format handled by this GPU is RGB565, what was causing artifacts in e.g. Rez (using ARGB1555) - this one I have fixed. There is one more issue in THPS2 when you try to exit to the main menu while playing - black screen appears (instead of "Loading...", like on Adreno GPUs).
> Anyway, these issues are not related to this pull request or the #1460.

MXIII-G - Android 5.1.1 - DDR3 2GB RAM+16GB 32Bits
CPU: Amlogic S812 Quad Core up to 2.0GHz (Cortex-A9)
GPU: Octa core ARM Mali-450 GPU up to 600MHz \ OpenGL 2.0 ES

The set of this box is good, I use it connected to a 32 "fullhd tv, and the resolution of the box in 1080p, but I tried others up to 480p and even modified the format the bug remain, interesting that in the game Mortal Kombat Gold this bug only appears in this palette, the rest is normal, so I have been testing in this box in emulators like parallel64, Mupenplus64 and PPSSPP, some games suffer with this bug in resolutions above 640x480, but if they are in resolution 320x240 or 480x360 the bug does not happen, it would be interesting to be able to set the reicast in 320x240, 480x360,640x480 .... but I do not know if it would be possible.

mar753 commented:
The bug you are describing is not caused by this PR, it may be hardware specific as I have mentioned or caused by something other. Report your bug by opening new thread in https://github.com/reicast/reicast-emulator/issues section.

When using code changes from this PR, we are able to lower the resolution in reicast experimental options (by lowering 'Resolution' percent seekbar). If your default output resolution is 1920x1080 (1080p), and you set this seekbar to e.g. 20%, your reicast resolution in every game will be: 20% * 1920x1080 what gives about 384x216 (without screen scissoring on both sides).

mar753 commented:
The bug you are describing is not caused by this PR, it may be hardware specific as I have mentioned or caused by something other. Report your bug by opening new thread in https://github.com/reicast/reicast-emulator/issues section.

When using code changes from this PR, we are able to lower the resolution in reicast experimental options on Android (by lowering 'Resolution' percent seekbar). If your default output resolution is 1920x1080 (1080p), and you set this seekbar to e.g. 20%, your reicast resolution in every game will be: 20% * 1920x1080 what gives about 384x216 (without screen scissoring on both sides).

baka0815 commented:
I tried running Crazy Taxy again and it's now running and I don't see any glitches on first sight.

baka0815 commented:
Here are some pictures under Linux with my AMD RX 580, Mesa 18.2:

Crazy Taxy at 20%
![ct-20-1](https://user-images.githubusercontent.com/15894566/51559075-45130680-1e81-11e9-93ac-32a338d2615d.png) ![ct-20-2](https://user-images.githubusercontent.com/15894566/51559077-46dcca00-1e81-11e9-8b08-6e45596ad60b.png)

Crazy Taxy at 33%
![ct-33-1](https://user-images.githubusercontent.com/15894566/51559096-4e9c6e80-1e81-11e9-8159-12bbf22a4919.png) ![ct-33-2](https://user-images.githubusercontent.com/15894566/51559101-4fcd9b80-1e81-11e9-922c-8c6249413305.png)

Crazy Taxy at 50%
![ct-50-1](https://user-images.githubusercontent.com/15894566/51559107-565c1300-1e81-11e9-8625-af2eed6c1271.png) ![ct-50-2](https://user-images.githubusercontent.com/15894566/51559110-578d4000-1e81-11e9-9341-59d90f3b47f8.png)

And for the fun, Crazy Taxy at 200%
![ct-200](https://user-images.githubusercontent.com/15894566/51559125-5fe57b00-1e81-11e9-91c9-7e39db6ba50c.png)

mar753 committed:
rend.ResolutionPercentage acceptable values are now from 1 to 100; naming fix

baka0815 commented:
Overall I think this looks good, but why stick to the x and y scale parameters if we're only having one setting now? I don't think we'll be scaling one axis differently than the other.

It would be great if you could change this last thing and then I'm fine merging this if no other dev objects (@dmiller423 @skmp @MrPsyMan)

mar753 commented:
Actually, x and y scale parameters are just leftovers from separate horizontal and vertical seekbars. I have just made one common seekbar (parameter) on Android that sets them both to the exact same value simultaneously (similarly when reading 'rend.ResolutionPercentage' from emu.cfg), so there is no difference at all. The benefit is that if we would like to make them separate in future (aspect ratio reasons), it will be easy. May be useful during debugging of some rendering issues as well.

baka0815 commented:
Well, I'm fine with that. Just waiting a couple of days to give the other devs the change to review before I merge it.

mar753 commented:
Sure, no problem

baka0815 commented:
@flyinghead I think he meant you 😉

baka0815 commented on core/nullDC.cpp:
```diff
@@ -299,6 +299,10 @@ void LoadSettings()
+	    VerticalResolution			= cfgLoadInt("config", "rend.ResolutionPercentage", 100);
+	settings.rend.
+	    HorizontalResolution		= cfgLoadInt("config", "rend.ResolutionPercentage", 100);
 
```
Maybe we should clamp the settings on other platforms (Linux) too:
```suggestion
	if (settings.rend.VerticalResolution <= 0 || settings.rend.VerticalResolution > 100)
	  settings.rend.VerticalResolution = 100;
	if (settings.rend.HorizontalResolution <= 0 || settings.rend.HorizontalResolution > 100)
	  settings.rend.HorizontalResolution = 100;
	  
```

mar753 replied:
Actually, clamping in the current solution will work for every GL/GLES platform, so no changes are needed

baka0815 commented:
And here are the German texts for `` reicast-emulator/shell/android-studio/reicast/src/main/res/values-de/strings.xml``:
````xml
    <string name="resolution">Auflösung (niedriger = schneller)</string>
    <string name="resolution_vertical">Vertikal</string>
    <string name="resolution_horizontal">Horizontal</string>
````

flyinghead commented:
This looks good to me

skmp commented:
My only concern is that rendeing to an off screen buffer and then to the
main screen can be much slower when no scaling happens. Otherwise, looks
fine. LTGM?

On Thu, 31 Jan 2019 at 16:53, flyinghead <notifications@github.com> wrote:

> This looks good to me
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/reicast/reicast-emulator/pull/1480#issuecomment-459395776>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAYAMqOzkqExLKaPyoxVzR8iF6Qs0H8kks5vIxFjgaJpZM4Z0drW>
> .
>
-- 
~skmp

mar753 committed:
German translation added

mar753 commented:
German translation added.

When no scaling happens we will not render to the off-screen buffer, everything will be as is now (rendering directly to the screen). Moreover performance difference in this case is rather negligible.

Emulatorer commented:
waiting for this :)

baka0815 commented:
Thanks @mar753 for your work on this!

verybadsoldier commented:
I bought a great new 4K TV recently and reicast now seems to insist on rendering games at native 4K resolution on my Intel NUC which looks great but also is very slow.
Am I right that I need this PR to have it running on a lower resolution or are there other options also?

skmp commented:
This is already merged, and then partially overwritten by flyinghead's changes. 

You should be able to use the slidebar in the video settings to reduce the internal resolution.

verybadsoldier commented:
Oh that's great, thank you! And sorry I didn't see that myself.

verybadsoldier commented:
One more question please: the scale factor does apply to the image width/height, right? So when having a scaling of 0.33 when using 1920x1080 natively then I get 640x480? I think I would like to have the original Dreamcast resolution.

Would be perfect if it would be possible to directly specify the target resolution (640x480) instead of the scaling factor.
